Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +17,66 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return currentAmpHours;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
-{
-    return false;
+{   
+    if( current < 0 )
+    {
+        return true;
+    }
+    else
+    {
+        return false;
+    }
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    double initialValueOfCharge = initialStateOfChargePercent_ * BATTERY_AMP_HOUR_CAPACITY; //initial state in miliseconds
+    double averageIncreaseValue = totalAmpHoursUsed() * 3.6e+6 ; //the increase in charge
+    if(averageIncreaseValue <= 0)
+    {
+        averageIncreaseValue = -averageIncreaseValue;
+    }
+    if( isCharging() )
+    {
+        // Ah / a
+        double batteryChargeRemaining = (averageIncreaseValue * 3.6e+6) * (BATTERY_AMP_HOUR_CAPACITY-initialValueOfCharge) / -current; //BatteryAMPHOUR should be totalAMpHour
+        double timeLeft = batteryChargeRemaining / averageIncreaseValue;
+        QTime timeUntilFinished = currentTime.addMSecs(timeLeft);
+        return timeUntilFinished;
+    }
+    else
+    {
+        //Ah x V / Watts
+        double timeLeft = ((averageIncreaseValue * voltage * 3.6e+6 * initialValueOfCharge) / -(voltage * current));    //AH value passed around
+        QTime timeUntilFinished = currentTime.addMSecs(timeLeft);
+        return timeUntilFinished;
+    }
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
-    // Hint: There are many different ways that the totalAmpHoursUsed can be updated
-    // i.e: Taking a running average of your data values, using most recent data points, etc.
+    {
+        voltage = batteryData.voltage;
+        current = batteryData.current; //get the current value
+        if(currentTime.toString() == NULL)
+        {
+            previousTime = QTime(0,0,0);
```

Comment:
You can't just set a time of 0, because the next time could be way in the future but that amount of time hasn't actually elapsed. You want to consider not changing the ampHours data when you only have one data point.

Setting a time of 0 for the first data point doesn't work, because that could create a huge time difference between the first data point and the previous time. This will greatly skew the calculations. One possibility is not changing the ampHours data when only one data point is available.